### PR TITLE
Basic postgres migration infrastructure

### PR DESCRIPTION
**Pull request by oetherington**
This is just very basic infrastruture, but it seemed like a good place to separate the PRs.

This main outcomes of this are `yarn migrate up` to run all outstanding migrations and `yarn migrate create --name=my-new-migration` to create a new migration. Down migations are currently disabled as we discussed, but this should be simple to change in the future if we want.

[Umzug](https://www.npmjs.com/package/umzug) seems pretty cool. It's actively developed and has almost 700,000 weekly downloads. It's got enough features to make life pretty straight-forward, but it's still low-level enough that it should fit in with our weird requirements.

I'm not totally convinced by the name `mongoMigrations` for the existing stuff. Other options could be `legacyMigrations` or `migrationScripts` maybe? I'm open to ideas.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203143496080116) by [Unito](https://www.unito.io)

### CODE REVIEW

**Review by Will-Howard — COMMENTED** [review]
Questions about this:
 - How is the ordering enforced, is it just alphabetical? What happens if you run `migrate up` and there is a pending out of order migration? I think it should fail in this case and make you sort out the ordering
 - What happens if one migration succeeds and the one after it fails, does it roll back both? (I think the answer is no, but that it should, we can (optionally) wrap the whole `up` command in a transaction to make this work)

Otherwise looks good though, nice and simple!

File: packages/lesswrong/server/migrations/2022.10.10T16.47.12.initial.ts
Comment (oetherington) [reply packages/lesswrong/server/migrations/2022.10.10T16.47.12.initial.ts:1]: Good call! The current behaviour is just the umzug default, but it looks like it's configurable.

File: package.json
Comment (oetherington) [reply package.json:268]: The main options that run directly in node were dbmate, marv, pg-migrate and db-migrate. They're all a bit higher level than umzug which makes them a little trickier to integrate how we want, plus they have varying levels of typescript support (not just with getting code compiled, but then with actually allowing us to import code from the rest of the codebase without ESM/commonjs conflicts).

There's also things like flyway and goose which run outside of node, but they also have the same typescript woes.

It seems the most popular way to do migrations in node these days is to use some big ORMish framework like knex or sequelize, but that felt like a commitment we don't want to make right now (and, very possibly, ever).

File: packages/lesswrong/server/migrations/2022.10.10T16.47.12.initial.ts
Comment (oetherington) [reply packages/lesswrong/server/migrations/2022.10.10T16.47.12.initial.ts]: Update - this was less configurable than expected. For now, I've created a fork to make it configurable, but I'll make a pull request and update the dependency when merged.

**Comment by oetherington** [comment]
@Will-Howard Thanks for reviewing!

WRT the two questions above:

- Right now, ordering is just enforced by a lexographical sort of the filenames (with timestamps) as you say. I'm not 100% sure changing this is necessary (I'm pretty sure most migration systems work this way?) but I could probably be convinced otherwise.
- Currently, there's actually no transactions (there definitely should be but I totally forgot). I'm not entirely convinced though that a later migration failing should cause an earlier successful migration to be rolled back, especially if we _do_ implement the strict ordering above. Again though, I'm open to being persuaded.

I feel like it would probably be a good idea to discuss this at some point as I'm not really feeling 100% confident about what the ideal solution is at the moment. I don't think these additions will change the base we already have here so it's probably safe to merge, and I'll add a separate ticket in Asana to deal with ordering and transactions.

The one problem holding up merging is that I had to use a custom version of umzug to change the timestamp format, and for some reason the github action doesn't like building the library. For now, I'll just reassign this PR to myself in Asana and remove the due date. Hopefully the umzug change will be merged quickly, otherwise I'll try to work out some temporary workaround.

**Comment by Will-Howard** [comment]
> * Right now, ordering is just enforced by a lexographical sort of the filenames (with timestamps) as you say. I'm not 100% sure changing this is necessary (I'm pretty sure most migration systems work this way?) but I could probably be convinced otherwise.

Ok this seems fine, I think there is a slight risk of migrations from different people being merged in out of order and causing problems, but I'm happy to wait and see if this happens. The last time I used migrations was with Django which declares specific dependencies in each migration, but maybe that was just overkill

> * Currently, there's actually no transactions (there definitely should be but I totally forgot). I'm not entirely convinced though that a later migration failing should cause an earlier successful migration to be rolled back, especially if we _do_ implement the strict ordering above. Again though, I'm open to being persuaded.

This is mainly so that if a migration fails during a deployment (which will fail the whole deployment) it reverts back to how it was before. The risk here is that you drop a column or something and the old code then doesn't work. Ofc people shouldn't be writing migrations that cause the old code to break because there is some time during the deployment where old code will be running against the migrated database anyway, so maybe this also isn't such a big deal. If it's easy to wrap the whole process in a transaction as part of what you do to wrap individual migrations then I would like to do this, if not then happy to leave it

Happy to merge this with just the umzug fix though, I'll approve it so you can merge as soon as that's done

**Review by Will-Howard — APPROVED** [review]

**Comment by oetherington** [comment]
@Will-Howard Both very good points. I'll think about this more before I start work on the implementation. Thanks! (And will merge this once umzug is ready!)

**Pull request merged**
